src/mscorlib/corefx/System/Runtime/InteropServices/StringBuffer.cs

```diff
@@ -2,129 +2,111 @@
         {
-            EnsureByteCapacity(minCapacity * (ulong)sizeof(char));
+            if (minCapacity > Capacity)
+            {
+                char[] oldBuffer = _buffer;
+                _buffer = ArrayPool<char>.Shared.Rent(minCapacity);
+                Array.Copy(oldBuffer, 0, _buffer, 0, oldBuffer.Length);
+                ArrayPool<char>.Shared.Return(oldBuffer);
+            }
         }
 
         /// <summary>
         /// The logical length of the buffer in characters. (Does not include the final null.) Will automatically attempt to increase capacity.
         /// This is where the usable data ends.
         /// </summary>
         /// <exception cref="OutOfMemoryException">Thrown if unable to allocate memory when setting.</exception>
-        /// <exception cref="ArgumentOutOfRangeException">Thrown if the set size in bytes is uint.MaxValue (as space is implicitly reserved for the trailing null).</exception>
-        public unsafe uint Length
+        /// <exception cref="ArgumentOutOfRangeException">Thrown if the set size in bytes is int.MaxValue (as space is implicitly reserved for the trailing null).</exception>
+        public int Length
         {
             get { return _length; }
             set
             {
-                if (value == uint.MaxValue) throw new ArgumentOutOfRangeException(nameof(Length));
-
                 // Null terminate
-                EnsureCharCapacity(value + 1);
-                CharPointer[value] = '\0';
+                EnsureCapacity(checked(value + 1));
+                _buffer[value] = '\0';
 
                 _length = value;
             }
         }
 
         /// <summary>
-        /// For use when the native api null terminates but doesn't return a length.
-        /// If no null is found, the length will not be changed.
-        /// </summary>
-        public unsafe void SetLengthToFirstNull()
-        {
-            char* buffer = CharPointer;
-            uint capacity = CharCapacity;
-            for (uint i = 0; i < capacity; i++)
-            {
-                if (buffer[i] == '\0')
-                {
-                    _length = i;
-                    break;
-                }
-            }
-        }
-
-        internal unsafe char* CharPointer
-        {
-            get
-            {
-                return (char*)VoidPointer;
-            }
-        }
-
-        /// <summary>
         /// True if the buffer contains the given character.
         /// </summary>
         public unsafe bool Contains(char value)
         {
-            char* start = CharPointer;
-            uint length = _length;
-
-            for (uint i = 0; i < length; i++)
+            fixed (char* start = _buffer)
             {
-                if (*start++ == value) return true;
+                int length = _length;
+                for (int i = 0; i < length; i++)
+                {
+                    if (start[i] == value) return true;
+                }
```

Comment:
Doesn't look like it would make a difference, actually; looks like the JIT will end up specializing ```GenericEqualityComparer<char>.IndexOf``` to be basically equivalent to this loop.  If that's true, is #6688 really the issue here?